website/docs/language/modules/develop/refactoring.html.md

```diff
@@ -0,0 +1,475 @@
+sections describe the various different refactoring use-cases, along with the
+appropriate addressing syntax for each situation.
+
+* [Renaming a Resource](#renaming-a-resource)
+* [Enabling `count` and `for_each` For a Resource](#enabling-count-and-for_each-for-a-resource)
+* [Renaming a Module Call](#renaming-a-module-call)
+* [Enabling `count` and `for_each` For a Module Call](#enabling-count-and-for_each-for-a-module-call)
+* [Splitting One Module into Multiple](#splitting-one-module-into-multiple)
+* [Removing `moved` blocks](#removing-moved-blocks)
+
+## Renaming a Resource
+
+The following is an expanded version of the example above showing the full
+configurations at each step.
+
+Imagine that the inital version of your module included a resource configuration
+like the following:
+
+```hcl
+resource "aws_instance" "a" {
+  count = 2
+
+  # (resource-type-specific configuration)
+}
+```
+
+Applying this configuration for the first time would cause Terraform to
+create `aws_instance.a[0]` and `aws_instance.a[1]`.
+
+If you later choose a better name for this resource, then you can change the
+name label in the `resource` block and record the old name inside a `moved` block:
+
+```hcl
+resource "aws_instance" "b" {
+  count = 2
+
+  # (resource-type-specific configuration)
+}
+
+moved {
+  from = aws_instance.a
+  to   = aws_instance.b
+}
+```
+
+When creating the next plan for each configuration using this module, Terraform
+will treat any existing objects belonging to `aws_instance.a` as if they had
+been created as `aws_instance.b`. `aws_instance.a[0]` will be treated as
+`aws_instance.b[0]`, and `aws_instance.a[1]` as `aws_instance.b[1]`.
+
+New instances of the module, which _never_ had an
+`aws_instance.a`, will ignore the `moved` block and propose to create
+`aws_instance.b[0]` and `aws_instance.b[1]` as normal.
+
+Both of the addresses in this example referred to a resource as a whole, and
+so Terraform recognizes the move for all instances of the resource. That is,
+it covers both `aws_instance.a[0]` and `aws_instance.a[1]` without the need
+to identify each one separately.
+
+Each resource type has a separate schema and so objects of different types
+are not compatible. Therefore although you can use `moved` to change the name
+of a resource, you _cannot_ use `moved` to change to a different resource type
+or to change a managed resource (a `resource` block) into a data resource
+(a `data` block).
+
+There is one exception to that constraint: a provider developer may mark a
+particular resource type name as deprecated and nominate a new name to replace
+it. In that case, the provider developer is responsible for making the old
+and new names schema-compatible, and can then tell Terraform that it's safe to
+support changing from the old to the new type via a `moved` block. Provider
+documentation will include an example for any situation where this exception
+applies.
```

Comment:
are we sure there will always be an example? I wonder if we should soften and say "should" or do we enforce this somehow?

Reply:
The mechanism I've described here doesn't exist yet today, so I wrote this with the intent that writing such documentation would be part of the work of making use of this new mechanism in providers.

However, it's true that this is written under the assumption that all such changes will be made by folks at HashiCorp who would therefore follow our documentation conventions/standards. We can't force third-party providers to have such documentation, so it is perhaps too strong to say "will".

With that said, there isn't really any way for a user to discover that this is possible _other than_ in documentation, so an undocumented instance of this would honestly be rather useless. I'm happy to change the wording as you suggested, but this is also a good note that we should make sure that the external provider-developer-facing documentation for this feature should mention this assumption too, once we have this facility available in the plugin SDK for provider developers to use. (This is actually not part of our initial scope of work here, so I could also just remove this statement altogether for now and restore it later once there's at least one resource type it's true for.)

Reply:
After re-reading the full section to recover the context, I decided to just remove this paragraph altogether for now.

We can add more words about this once this mechanism is implemented and so it'll presumably be clearer to us what provider developers will be expected to do in order to enable it.